### PR TITLE
Refactor/loaders

### PR DESCRIPTION
**Pull request by collindutter**
- [x] I have read and agree to the contributing guidelines for [submitting new pull requests](https://github.com/griptape-ai/griptape?tab=readme-ov-file#submitting-pull-requests).

### Added
- `BaseFileLoader` for Loaders that load from a path.
- `BaseLoader.fetch()` method for fetching data from a source.
- `BaseLoader.parse()` method for parsing fetched data.
- `BaseFileManager.encoding` to specify the encoding when loading and saving files.
- `BaseWebScraperDriver.extract_page()` method for extracting data from an already scraped web page.
- `TextLoaderRetrievalRagModule.chunker` for specifying the chunking strategy.
- `file_utils.get_mime_type` utility for getting the MIME type of a file.

### Changed
- **BREAKING**: Removed `BaseFileManager.default_loader` and `BaseFileManager.loaders`.
- **BREAKING**: Loaders no longer chunk data, use a Chunker to chunk the data.
- **BREAKING**: Removed `fileutils.load_file` and `fileutils.load_files`.
- **BREAKING**: Removed `loaders-dataframe` and `loaders-audio` extras as they are no longer needed.
- **BREKING**: `TextLoader`, `PdfLoader`, `ImageLoader`, and `AudioLoader` now take a `str | PathLike` instead of `bytes`.
- **BREAKING**: Removed `DataframeLoader`.
- `LocalFileManagerDriver.workdir` is now optional.
- `filetype` is now a core dependency.
- `FileManagerTool` now uses `filetype` for more accurate file type detection.
- `BaseFileLoader.load_file()` will now either return a `TextArtifact` or a `BlobArtifact` depending on whether `BaseFileManager.encoding` is set.

The purpose of this PR was to clean up the Loader interface, and their define purpose. 

Loaders fetch data from a source, and parse it into Artifacts. Loaders do _not_ chunk data, that is the role of Chunkers.

We provide 4 top level Loaders that provide from a variety of sources:
- BaseFileLoader: Loads data from a file.
- WebLoader: Loads data from a web page.
- SqlLoader: Loads data from a SQL database.
- EmailLoader: Loads data from an email inbox.
  

BaseFileLoader then has subclasses that provide file-type specific parsing logic:
  - TextLoader
  - ImageLoader
  - AudioLoader
  - CsvLoader
  - BlobLoader
In a future PR, I'd like this parsing logic to live in a new class of Driver, File Parser Driver(?), instead of Loaders.
## Issue ticket number and link
Closes https://github.com/griptape-ai/griptape/issues/1102

----
📚 Documentation preview 📚: https://griptape--1116.org.readthedocs.build//1116/

### CODE REVIEW

**Comment by collindutter** [comment]
Note that this PR's base branch is `refactor/artifacts` so no tests have run yet.

File: griptape/loaders/base_loader.py [hunk 1]
```diff
@@ -39,10 +56,8 @@ def load_collection(
             },
         )
 
-    def to_key(self, source: Any, *args, **kwargs) -> str:
+    def to_key(self, source: Any) -> str:
```
Comment (dylanholmes) [inline griptape/loaders/base_loader.py:59]: `source: S`?

File: griptape/loaders/email_loader.py [hunk 2]
```diff
@@ -11,7 +11,7 @@
 
 
 @define
-class EmailLoader(BaseLoader):
+class EmailLoader(BaseLoader["EmailLoader.EmailQuery", list[bytes], ListArtifact]):  # pyright: ignore[reportGeneralTypeIssues]
     @define(frozen=True)
     class EmailQuery:
```
Comment (dylanholmes) [inline griptape/loaders/email_loader.py:16]: I know we like the nested classes, but wouldn't it be easier to just allow related classes in the same file?

The choice for nested classes was really about style after all wasn't it?

File: docs/examples/src/talk_to_a_pdf_1.py [hunk 3]
```diff
@@ -30,9 +31,10 @@
     rag_engine=engine,
 )
 
-artifacts = PdfLoader().load(response.content)
+artifacts = PdfLoader().parse(response.content)
+chunks = TextChunker().chunk(artifacts)
 
-vector_store.upsert_text_artifacts({namespace: artifacts})
+vector_store.upsert_text_artifacts({namespace: chunks})
```
Comment (dylanholmes) [inline docs/examples/src/talk_to_a_pdf_1.py:37]: The method name looks a little awkward now, what do you think?

File: docs/examples/src/query_webpage_1.py [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
 from griptape.loaders import WebLoader
 
 vector_store = LocalVectorStoreDriver(embedding_driver=OpenAiEmbeddingDriver(api_key=os.environ["OPENAI_API_KEY"]))
 
-artifacts = WebLoader(max_tokens=100).load("https://www.griptape.ai")
+artifacts = WebLoader().load("https://www.griptape.ai")
+chunks = TextChunker().chunk(artifacts)
 
-for a in artifacts:
+for a in chunks:
```
Comment (dylanholmes) [inline docs/examples/src/query_webpage_1.py:12]: `for c...`?

File: griptape/loaders/base_loader.py [hunk 5]
```diff
@@ -1,33 +1,50 @@
...
+        artifact.reference = self.reference
+
+        return artifact
+
+    @abstractmethod
+    def fetch(self, source: S) -> F: ...
 
     @abstractmethod
-    def load(self, source: Any, *args, **kwargs) -> BaseArtifact | Sequence[BaseArtifact]: ...
+    def parse(self, data: F) -> A: ...
```
Comment (dylanholmes) [inline griptape/loaders/base_loader.py:40]: `convert` or `to_artifact` might describe the behavior a little better?

File: griptape/loaders/base_loader.py [hunk 6]
```diff
@@ -1,33 +1,50 @@
...
 
 if TYPE_CHECKING:
-    from collections.abc import Mapping, Sequence
+    from collections.abc import Mapping
 
-    from griptape.artifacts import BaseArtifact
+    from griptape.common import Reference
+
+S = TypeVar("S")  # Type for the input source
+F = TypeVar("F")  # Type for the fetched data
```
Comment (dylanholmes) [inline griptape/loaders/base_loader.py:19]: Couple more options: Request/Response, Input/Output

**Review by dylanholmes — DISMISSED** [review]
The examples look so much cleaner! 🫧 

All the comments are minor or optional.

File: docs/examples/src/talk_to_a_pdf_1.py (hunk 3 above)
Comment (collindutter) [reply docs/examples/src/talk_to_a_pdf_1.py:37]: I don't think so. Artifacts are just data, chunks are a type of data. I could rename `chunks` back to `artifacts` but there might be times when you want to upload non-chunk Artifacts.

**Review by dylanholmes — DISMISSED** [review]

**Review by dylanholmes — APPROVED** [review]

**Pull request merged**